src/lib/Occurrence.hs

```diff
@@ -146,24 +160,40 @@ data Access (n::S) where
   -- Corresponds to `for`.
   ForEach :: forall n l. (NameBinder AtomNameC n l) -> Access l -> Access n
 
--- These are the indexing expressions we model.
+deriving instance Show (Access n)
+
+accessOnce :: Access n
+accessOnce = Location [] One
+
+accessMuch :: Access n
+accessMuch = Location [] Unbounded
+
+-- Prepend the given indexing expression onto an existing Access
+location :: IxExpr n -> Access n -> Access n
+location ix (Location ixs a) = Location (ix:ixs) a
+location _ _ = error "Cannot nest general Access in Location"
+
+-- These are the indexing expressions we model.  An IxExpr n is an abstraction
+-- of a function from names in the n scope, to either a single index location or
+-- to a set of index locations.  We only track the operations we have a hope of
+-- being able to analyze usefully.
 data IxExpr (n::S) =
   -- Indexing by an identifier in scope
     Var (Name AtomNameC n)
   -- An index of product type
   | Product [IxExpr n]
   -- An index of sum type
   | Inject Int (IxExpr n)
-  -- Unknown function of the given names
-  | Unknown [Name AtomNameC n]
-  -- All indices are accessed.  This constructor only appears after
-  -- `iterate`, below.
+  -- Unknown deterministic function of the given names
+  | Deterministic [Name AtomNameC n]
+  -- All indices are accessed, or we are conservatively assuming they are
+  -- because we don't know which are accessed.
```

Comment:
Is this form really sufficient? I would expect that we might want to distinguish between "all indices" coming form `for` loops and "any function" since "all indices" implies that every index is used once, while "any function" can do anything, including accessing an index twice?

Reply:
Yes, it is sufficient.  Explanation added.